### PR TITLE
126 influence analysis

### LINKED ISSUE #126: Influence analysis using topic modelling

### PR DESCRIPTION
**Pull request by Juan-Mateos**
Closes #126 

This includes:
* Changes in `requirements.txt`
* A summary of the method and how to implement it [here](https://github.com/nestauk/ai_genomics/tree/126_influence_analysis/ai_genomics/analysis#openalex----analysis-of-influence)
* `ai_genomics/analysis/influence/openalex_influence_analysis.py`, the script that implements the analysis above (you can also use it as a rough tutorial for the topic modelling utilities below).
* `ai_genomics/utils/text_modeling.py`, a script for text processing and _fast_ topic modelling.
* Changes in `ai_genomics/getters/openalex.py` to fetch the influence results from S3.

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

File: ai_genomics/getters/openalex.py [hunk 1]
```diff
@@ -141,22 +141,61 @@ def work_abstracts(discipline: str, years: List) -> Dict:
...
+
+    return ai_genom_getter("openalex_works", "csv", local)
+
+
+def get_openalex_ai_genomics_concepts(local: bool = True) -> pd.DataFrame:
+    """Returns dataframe of AI in genomics OpenAlex concepts"""
+
+    return ai_genom_getter("openalex_concepts", "csv", local)
 
-def get_openalex_entities() -> Mapping[str, Mapping[str, Union[str, str]]]:
```
Comment (india-kerle) [inline ai_genomics/getters/openalex.py:157]: I'd keep this!

File: ai_genomics/analysis/README.md [hunk 2]
```diff
@@ -31,3 +31,14 @@ Run `python ai_genomics/analysis/ai_genomics_experts/gtr_people.py` to generate
...
 You can run change the `reproduce` parameter to re-run the analysis from scratch. This includes creating vector representations of all sampled projects (which takes around 1hr locally).
+
+## OpenAlex -- analysis of influence
+
+Run `python ai_genomics/analysis/influence/openalex_influence_analysis.py` to perform a semantic analysis of discipline influence in our corpus of AI genomics research. This works as follows:
+
+1. Train topic model on corpus including AI genomics, AI, and Genomics research
+2. Calculate a discipline weight that captures the share of activity in a topic accounted by AI or Genomics e.g. if 80% of the top documents in a topic are in genomics, then it's weight is 0.8 (and the AI weight is 0.2)
+3. Calculate overall influence of a discipline (AI or Genomics) in a document by weighting each topic weight in a document by its discipline weight and aggregating them over disciplines
+4. The script validates findings by checking publication venues for documents in different positions of the "influenced by AI" distribution.
```
Comment (india-kerle) [inline ai_genomics/analysis/README.md:42]: a really nice way to evaluate without any manual intervention!

File: ai_genomics/analysis/influence/openalex_influence_analysis.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+NUM_TOPICS = 100
+
+if __name__ == "__main__":
+
+    logging.info("Reading data")
+    works = oal.get_openalex_ai_genomics_works()
+    abstrs = oal.get_openalex_ai_genomics_abstracts(local=False)
+
+    logging.info("Sampling")
+    # Sample all AI genomics and a subset of AI and genomics work
```
Comment (india-kerle) [inline ai_genomics/analysis/influence/openalex_influence_analysis.py:35]: given that we're sampling a lot, might be nice to save a sample of ids in s3 so we don't have to load all the works and then sample every time

File: ai_genomics/analysis/influence/openalex_influence_analysis.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    # Get the ids that belong to each category
+    ai_gen_id, ai_id, gen_id = [
+        set(works_sampled.loc[works_sampled[var] == True]["work_id"])
+        for var in ["ai_genomics", "ai", "genomics"]
+    ]
+
+    logging.info("Text processing and topic modelling")
+
+    # This is the prototype text pipeline. It could be refactored
+    # into spacy pipelines or something
```
Comment (india-kerle) [inline ai_genomics/analysis/influence/openalex_influence_analysis.py:62]: I reckon that's fine! We could eventually pull it into a text_preprocessing utils file

File: ai_genomics/analysis/influence/openalex_influence_analysis.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    logging.info("Validation of results")
+    logging.info("Do the topic-disc-weighted scores match the category?")
+    # Mean discipline scores for papers in different categories
+    logging.info(
+        paper_disc_score.groupby(["cat_assigned", "topic_category"])[
+            "weighted_disc_score"
+        ].mean()
+    )
+
+    logging.info(
```
Comment (india-kerle) [inline ai_genomics/analysis/influence/openalex_influence_analysis.py:147]: nice!

File: ai_genomics/analysis/influence/openalex_influence_analysis.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        ]
+
+        logging.info(
+            results_meta["venue_display_name"].value_counts(normalize=True).head(n=10)
+        )
+
+        logging.info("\n")
+
+    logging.info("Saving provisional results")
+    save_to_s3(
```
Comment (india-kerle) [inline ai_genomics/analysis/influence/openalex_influence_analysis.py:181]: nice, i've checked the output and looks good

File: ai_genomics/analysis/README.md [hunk 7]
```diff
@@ -31,3 +31,14 @@ Run `python ai_genomics/analysis/ai_genomics_experts/gtr_people.py` to generate
 Run `python ai_genomics/analysis/gtr/gtr_cluster_analysis.py` to reproduce the prototype cluster analysis of GtR clusters. Note that this analysis uses the same sampled dataset as reported by JMG.
 
 You can run change the `reproduce` parameter to re-run the analysis from scratch. This includes creating vector representations of all sampled projects (which takes around 1hr locally).
+
+## OpenAlex -- analysis of influence
+
+Run `python ai_genomics/analysis/influence/openalex_influence_analysis.py` to perform a semantic analysis of discipline influence in our corpus of AI genomics research. This works as follows:
```
Comment (india-kerle) [inline ai_genomics/analysis/README.md:37]: I was able to run the script but I initially had some environment issues due to spacy annoyingly - 

I was ultimately able to solve by conda installing instead:
`conda install -c conda-forge spacy
`

**Review by india-kerle — COMMENTED** [review]
Thanks Juan for tagging me! I ran your script and checked your output - looks great. I had issues with spacy that I solved by conda installing instead i.e. `conda install -c conda-forge spacy`

I think its a really interesting analysis - just thinking about what ada could be excited about with this and what we haven’t looked into much. Now that we have a sliding scale of AI influence at a work id level, we could aggregate or have some mean AI influence score over time at the venue/institution level and report on if i.e. traditionally genomics only venues have published more AI influenced papers over time

File: ai_genomics/getters/openalex.py (hunk 1 above)
Comment (Juan-Mateos) [reply ai_genomics/getters/openalex.py:157]: +1 added it again

File: ai_genomics/analysis/README.md (hunk 2 above)
Comment (Juan-Mateos) [reply ai_genomics/analysis/README.md:42]: 😃

File: ai_genomics/analysis/influence/openalex_influence_analysis.py (hunk 3 above)
Comment (Juan-Mateos) [reply ai_genomics/analysis/influence/openalex_influence_analysis.py:35]: That's a great point although in this script we would need to read the works eventually in order to perform the validation so I will leave it as is for now.

File: ai_genomics/analysis/influence/openalex_influence_analysis.py (hunk 4 above)
Comment (Juan-Mateos) [reply ai_genomics/analysis/influence/openalex_influence_analysis.py:62]: +1

File: ai_genomics/analysis/influence/openalex_influence_analysis.py (hunk 6 above)
Comment (Juan-Mateos) [reply ai_genomics/analysis/influence/openalex_influence_analysis.py:181]: +1

File: ai_genomics/analysis/influence/openalex_influence_analysis.py (hunk 5 above)
Comment (Juan-Mateos) [reply ai_genomics/analysis/influence/openalex_influence_analysis.py:147]: 😃

**Comment by Juan-Mateos** [comment]
Thank you for your comments! I have... 
* put `spacy` in `environment.yaml` and removed it from `requirements.txt`
* brought back `get_openalex_entities`

I explain above why I haven't implemented an ids getter (other than laziness :-D) but the idea is good and might come up down the line.

**Comment by india-kerle** [comment]
great! thanks for addressing those comments - feel free to merge.

**Pull request merged**